src/engine/renderer/tr_shade.cpp

```diff
@@ -903,46 +882,184 @@ static void Render_lightMapping( shaderStage_t *pStage )
 			break;
 	}
 
-	// u_LightMap, u_DeluxeMap
-	image_t *lightmap;
-	image_t *deluxemap;
+	enum class Light { FullBright, Vertex, Grid, Map };
+	Light light = Light( r_lighting->integer );
 
-	if ( noLightMap )
+	enum class Deluxe { None, Grid, Map };
+	Deluxe deluxe = Deluxe::None;
+
+	if ( light != Light::FullBright )
 	{
-		lightmap = tr.whiteImage;
-		enableLightMapping = true;
-	}
-	else if ( enableLightMapping )
+		/* TODO: investigate what this is. It's probably a hack to detect some
+		specific use case. Without knowing which use case this takes care about,
+		any change in the following code may break it. Or it may be a hack we
+		should drop if it is for a bug we don't have anymore. */
+		bool hack = tess.numSurfaceStages > 0 && tess.surfaceStages[0]->rgbGen == colorGen_t::CGEN_VERTEX;
+
+		// Also use fullbright on “surfaceparm nolightmap” materials.
+		if ( ( tess.surfaceShader->surfaceFlags & SURF_NOLIGHTMAP ) && !hack )
+		{
+			light = Light::FullBright;
+		}
+	}	
+
+	// u_Map, u_DeluxeMap
+	image_t *lightmap = tr.whiteImage;
+	image_t *deluxemap = tr.whiteImage;
+
+	uint32_t stateBits = pStage->stateBits;
+
+	if ( light == Light::Map )
 	{
-		lightmap = GetLightMap();
+		if ( tess.bspSurface )
+		{
+			/* TODO: There is no reason this test fails:
+			( pStage->implicitLightmap || pStage->type == stageType_t::ST_LIGHTMAP )
+			Anyway, this test is useless. */
+			bool hasMap = tr.worldLightMapping
+				&& ( tess.lightmapNum >= 0 && tess.lightmapNum <= tr.lightmaps.currentElements );
+
+			if ( !hasMap )
+			{
+				/* Use vertex light instead of grid light as a fallback on world
```

Comment:
Well, both are true, but the one I wrote lacks context. There are no lightmap, but two remaining source of lights, vertex light or grid light. As a fallback to the lack of lightmap, use vertex light and not grid light. Using vertex light instead of grid light in such case is one of the bugfix this PR brings.

Reply:
I guess I can reword that comment.

Reply:
Reworded.